Create a blog post describing the Kyma website (#573)
kyma-project/website: content/blog-posts/2020-05-19-kyma-website/index.md

---
title: "Tools, sources, and all the magic behind the Kyma website"
author:
  name: "Maciej Urbanczyk and Karolina Zydek"
tags:
  - graphql
  - netlify
  - gatsby
redirectFrom:
  - "/blog/2020-05-19-kyma-website"
---

Our website, just like success, has many fathers maintaining it from the very first day it was born out of the burning need for a homepage for our soon-to-be-open-sourced Kyma. It evolved in time, in an agile spirit, extended with new views and features that were added when the need arose or an idea for improvement popped into our heads. As much as we love it, we realize it might seem a bit complex, especially for those who would like to contribute to it for the first time. In this post we explain which tools we decided to use, how the website is built, and where all the sources sit - all this aiming to "tame the beast" and bring it a bit closer to you.

<!-- overview -->

## Building tool

When it comes to the tool we wanted to build our website with, we decided to choose a static site generator that would:
- Make the development process easier by providing a set of ready-to-use plugins and templates.

> - Provide support for Search Engine Optimization (SEO), increasing our website's visibility on search engines.

Author: I mentioned SPA briefly, have a look if that is enough :)

## Deployment tool

Although we used GitHub pages initially to deploy and host our website, we turned to [Netlify](https://www.netlify.com/) that offers a favorable [policy plan](https://www.netlify.com/legal/open-source-policy/) for the open-source projects.

As the website expanded, we simply needed a platform that would meet the needs ahead of us. Netlify managed to do so by offering:

- [Continuous deployment](https://www.netlify.com/blog/2015/09/17/continuous-deployment/) enabling automatic website built triggered by GitHub commits.

- [Deploy previews](https://www.netlify.com/blog/2016/07/20/introducing-deploy-previews-in-netlify/) built on pull requests, allowing you to detect any rendering issues before they are merged to the master.

- [AWS functions](https://docs.netlify.com/functions/overview/) for running on-demand code triggered by events coming from the GitHub and ZenHub APIs. In Kyma, we use it as a trigger for running the build and deployment of given website views. This function is triggered whenever you:

>   - Merge content to the `docs` folder in the `kyma` repository, to the `community` repository, or to the `content` folder in the `website` repository

Author: I simply removed info on website.

  - Create a release in the `kyma` repository. This triggers the rebuild of the **Docs** view and triggers a new release switcher.

  ![Release switcher](./release-switcher.png)

>   - Add or remove the `Epic` label to issues (ZenHub API), the release for which the epic is planned (ZenHub API), the issue's capability label (GitHub API), or the issue title.

Author: (Specify that any change triggers the build)

  ![Issue labels](./issue-labels.png)

  - Change ClusterAssetGroups that provide the structure for the left topic navigation in **Docs** and **Community**.

  ![Left navigation in docs](./left-navigation.png)

## Custom rendering

To customize content rendering in some of the views, we put our own twists on the standard logic provided by Gatsby.

### Documentation component

> In Kyma, we render the documentation sources both on the website under the **Docs** view and as built-in documentation in the UI on every Kyma cluster. To unify the way in which it is shown in both places, we created our own React [documentation component](https://github.com/kyma-incubator/documentation-component) for rendering such specification formats as Markdown, OpenAPI, AsyncAPI, and OData. We use this component on the website to render Markdown and OpenAPI specifications with [custom styles](https://github.com/kyma-project/website/tree/master/src/components/generic-documentation/render-engines) applied in chosen views. The documentation component also provides the documentation toggle, right navigation and scrollspy in **Docs** and **Community** views:

Author: documentation toggle - tabs ;)

> ![Features provided by the documentation component](./documentation-component.png)

Author: Clarified that we both meant tabs :)

### React-markdown library

> We also wanted to customize the way we render some Markdown elements. Since Gatsby didn't support that, and we couldn't find any other tool for that, we created the [react-markdown library](https://github.com/kyma-incubator/documentation-component/tree/master/packages/markdown-render-engine) that is our wrapper for the [`react-markdown`](https://github.com/rexxars/react-markdown). We used it in **Docs**, **Blog**, and **Community** views to customize such Markdown elements as panels, or icons next to external links.

Reviewer: Gatsby support custom rendering Markdown elements, but it has less functionality than react-markdown and we also have, as you mentioned, docs view in Kyma Console UI, so due to this two problems, we decide to use custom rendering by `react-markdown` :)

Author: (practical reasons)

![ClusterAssetGroups for the left navigation](./ClusterAssetGroups.png)

## Views and their sources

![Views on the website](./views.png)

Each view on the website takes its sources from a different repository.

- **Docs** - `docs` folder in the `kyma` repository

> - **Blog** - `blog-post` folder in the `website` repository

Author: I added them originally but then I thought it would be too much as they are already in the contributing section. I didn't want to pollute the text with so many links.

- **Community** - `community` repository
- **Roadmap** - `capabilities` folder in the community repository for descriptions of our project areas, and GitHub issues with `Epic` and a given capability's labels for the roadmap details.
- **Landing page** - `content` folder in the `website` repository, including the banner, and the **Used by** section with Kyma users

Before the website build, all this content is moved to the `website` repository by the [content loader](https://github.com/kyma-project/website/tree/master/tools/content-loader) - our in-house TypeScript tool we use for fetching:
- Content from various repository sources
- Details of issues from ZenHub and GitHub APIs

## Building and deployment process

Now that you know all pieces of the puzzle, let's have a look at how they fit together. The diagram and description beneath show the whole website building process triggered after merging a pull request.

> ![Website building flow](./building-process.svg)

Reviewer: Can we change this diagram? I would point more precisely that Gatsby create the Graphql Schema and then build static HTML site, because already when I see diagram I come to the conclusion that the content loader is preparing this scheme and this is not true.

1. When you merge your PR to the `kyma/docs`, `website` or `community` repositories, the GitHub API sends an event to the Netlify function that triggers the master build.

> 2. The build triggers the content loader to fetch the given repository content, along with related ClusterAssetGroups, and perform initial data serialization and filtering. The content loader then moves the fetched content into the temporary folders in the `website` repository.

Author: (remove temporary)

3. Gatsby reads this content through the connected plugins and retrieves selected data, such as metadata in docs. It later transforms it into a GraphQL schema that could be pulled by React components.

4. Gatsby uses the [Browser API](https://www.gatsbyjs.org/docs/browser-apis/) to build particular static HTLM sites from React components, using data from the GraphQL schema.

> 5. Gatsby performs the optimization process - it compiles the source code so that the sites could load faster on all sorts of devices.

Reviewer: I have a very big problem with that point :D We can point here to some Gatsby docs about his performance, but docs are scattered around various places in the documentation and I don't know if there is any sense for that... Also  sentence `...it compiles the source code so that the sites could load faster on all sorts of devices.` we can remove. Info about optimization is enough.

Reviewer: We can use this doc https://www.gatsbyjs.org/docs/performance/

Author: (Just "Gatsby performs the optimization process " and add the link)

| **Landing page** | [`website/content/banner/slides.yml`](https://github.com/kyma-project/website/blob/master/content/banner/slides.yml) | Banner | Follow this [instruction](https://github.com/kyma-project/website/blob/master/docs/banner-modification.md). |
| **Docs** | [`kyma/docs`](https://github.com/kyma-project/kyma/tree/master/docs) | Document or topic | Add a document that follows one of the [templates](https://kyma-project.io/community/guidelines/templates/#document-types-templates-document-types-templates-document-types-for-kyma-components) or follow instructions to [add a new topic](https://kyma-project.io/community/guidelines/content/#add-new-documentation-to-the-website-add-new-documentation-to-the-website). |
| **Blogs** | [`website/content/blog-post`](https://github.com/kyma-project/website/tree/master/content/blog-posts) | Blog post | Follow the [instruction](https://github.com/kyma-project/website/blob/master/docs/write-blog-posts.md). |
| **Community** | [`community`](https://github.com/kyma-project/community) | Document | |

> | **Roadmap** | All repos | Epic | Log and issue, assign it to the proper milestone on Zenhub, and add your capability's and `epic` labels. |

Reviewer: `Epic` label :P

---
title: "Tools, sources, and all the magic behind the Kyma website"
author:
  name: "Maciej Urbańczyk and Karolina Zydek"
tags:
  - graphql
  - netlify
  - gatsby
  - community
redirectFrom:
  - "/blog/2020-05-19-kyma-website"
---

> Our website, has many "fathers" who have maintained it from the very first day it was born out of the burning need for a homepage for our soon-to-be-open-sourced Kyma. It has evolved in time, in an agile spirit, extended with new views and features that were added whenever the need arose or an idea for improvement popped into our heads. As much as we love it, we realize it might seem a bit complex, especially for those who contribute to it for the first time. In this post we explain which tools we decided to use, how the website is built, and where all the sources sit - all this aiming to "tame the beast" and bring it a bit closer to you.

Author: I changed it to `from the day it was born` as I didn't catch the difference in the second option. But your 1st suggestion is ok.

> Our website, has many "fathers" who have maintained it from the very first day it was born out of the burning need for a homepage for our soon-to-be-open-sourced Kyma. It has evolved in time, in an agile spirit, extended with new views and features that were added whenever the need arose or an idea for improvement popped into our heads. As much as we love it, we realize it might seem a bit complex, especially for those who contribute to it for the first time. In this post we explain which tools we decided to use, how the website is built, and where all the sources sit - all this aiming to "tame the beast" and bring it a bit closer to you.

Author: or maybe `ever since it was born` - is that ok?

> Our website has many "fathers" who have maintained it ever since it was born out of the burning need for a homepage for our soon-to-be-open-sourced Kyma. It has evolved in time, in an agile spirit, extended with new views and features that were added whenever the need arose or an idea for improvement popped into our heads. As much as we love it, we realize it might seem a bit complex, especially for those who contribute to it for the first time. That's why we use this post to explain which tools we decided to use, how the website is built, and where all the sources sit — all this aiming to "tame the beast" and bring it a bit closer to you.

Reviewer 2: just for consideration `and where all the sources sit — all this aiming to "tame the beast" and bring it a bit closer to you.` if it's about bringing the beast closer, then it's fine :) if about the idea then I would change `it` to state it explicitly.

Author: beast = website, so I guess that can stay as it. is :)

<!-- overview -->

## Building tool

When it comes to the tool we wanted to build our website with, we decided to choose a static site generator over a single-page application (SPA) to:
- Make the development process easier by providing a set of ready-to-use plugins and templates.
- Provide support for search engine optimization (SEO), increasing our website's visibility in search results.
- Ensure quick load time.

> Our choice fell on [Gatsby](https://www.gatsbyjs.org/), partially because it is based on React that our frontend developers were used to and particularly fond of. Gatsby also uses GraphQL to query and pull source data, and we already used it for our [Console Backend Service](https://kyma-project.io/docs/components/console/#details-console-backend-service) in Kyma for the communication between Console UI views and k8s resources.

Author: I changed `partially` to `partly` but I'm not sure about the non-restrictive clause here - that info refers directly to React and basically justifies its choice.

> In its initial version, the website was quite simple and almost solely based on the logic provided by Gatsby. Basically, we wrote React components for particular views following Gatsby's tutorials and templates, used the available plugins to extend the data and content. Gatsby then pulled all the data from specified Markdown sources, built, and rendered the content on GitHub pages.

Author: I change to `defined `then :)

## Deployment tool

Although we used GitHub pages initially to deploy and host our website, we turned to [Netlify](https://www.netlify.com/) that offers a favorable [policy plan](https://www.netlify.com/legal/open-source-policy/) for the open-source projects.

> As the website expanded, we simply needed a platform that would meet all the needs ahead of us. Netlify managed to do so by offering:

Author: `that would help us to tackle the challenges ahead of us`?

> - [Continuous deployment](https://www.netlify.com/blog/2015/09/17/continuous-deployment/) enabling automatic website built triggered by GitHub commits to the `master` branch. With GitHub pages, we used CI from Prow that we had to maintain ourselves, so we gladly moved to the one offered by Netlify.

Author: maybe sth like this?
`[Continuous deployment](https://www.netlify.com/blog/2015/09/17/continuous-deployment/) thanks to which the website is built automatically every time you commit changes to the `master` branch`

> - [Deploy previews](https://www.netlify.com/blog/2016/07/20/introducing-deploy-previews-in-netlify/) built on pull requests (PRs), allowing you to detect any rendering issues before they are merged to the `master` branch.

Author: Done :)

- [AWS functions](https://docs.netlify.com/functions/overview/) for running on-demand code triggered by events coming from the GitHub API. In Kyma, we use them as a trigger for running the build and deployment of specific website views. Such a function is triggered whenever you:

>   - Merge content to the `docs` folder in the `kyma` repository and to the `community` repository.

Author: Yeah, I was thinking how to approach links in the post in general and decided that the table at the end with all the links would be enough - otherwise, the post would be cluttered with links and I wanted to avoid that.

  - Create a release in the `kyma` repository. This triggers the full website rebuild and adds a new release switcher to the **Docs** view.

  ![Release switcher](./release-switcher.png)

  - Modify issues with the `Epic` label. For example, when you add or remove capability labels or the release for which the epic is planned.

  ![Issue labels](./issue-labels.png)

  - Change the `.yaml` files that provide the structure for the left-side navigation in **Docs** and **Community**. You can read more about them in the [ClusterAssetGroups](#cluster-asset-groups) section.

>   ![Left-side navigation in docs](./left-side-navigation.png)

Reviewer 2: This is rendered as text on the website 🤔 
![2020-05-28_09-06-58](https://user-images.githubusercontent.com/17266942/83110296-2222e300-a0c3-11ea-9b0b-3ca0ad1cc3ad.png)

Author: yup, wrong name of the screenshot, fixed that :)

>   ![Left-side navigation in docs](./left-side-navigation.png)

Author: Fixed :)

>   ![Left-side navigation in docs](./left-side-navigation.png)

Author: Left as it is, just changed the image name :)

## Custom rendering

To customize content rendering in some of the views, we put our own twists on the standard logic provided by Gatsby.

### Documentation component

In Kyma, we render the documentation sources both on the website under the **Docs** view and as built-in documentation in the UI on every Kyma cluster. To unify the way in which it is displayed in both places, we created our own React [documentation component](https://github.com/kyma-incubator/documentation-component) to render Markdown formats and specifications such as OpenAPI, AsyncAPI, and OData. We use this component on the website to render Markdown formats and OpenAPI specifications with [custom styles](https://github.com/kyma-project/website/tree/master/src/components/generic-documentation/render-engines) applied in chosen views. The documentation component also provides tabs, copy buttons next to code snippets, and scrollspy in the **Docs** and **Community** views:

![Features provided by the documentation component](./documentation-component.png)

### React-markdown library

> We also wanted to customize the way we render some Markdown elements. Since Gatsby couldn't cater for all our needs, and we couldn't find any other suitable tool, we created one on our own. We needed a unified library that we could use for both the website and the cluster documentation. We came up with the [react-markdown library](https://github.com/kyma-incubator/documentation-component/tree/master/packages/markdown-render-engine) as a wrapper for [`react-markdown`](https://github.com/rexxars/react-markdown). We used it in the **Docs**, **Blog**, and **Community** views to customize such Markdown elements as panels and icons that appear next to external links.

Author: hmm actually that's its name but I decided to write it differently to avoid this dilemma here:
```
## Custom rendering
### In-house documentation component
### New react-markdown library
```

We also wanted to customize the way we render some Markdown elements. Since Gatsby couldn't cater for all our needs, and we couldn't find any other suitable tool, we created one on our own. We needed a unified library that could be used both on the website and for the cluster documentation. We came up with the [react-markdown library](https://github.com/kyma-incubator/documentation-component/tree/master/packages/markdown-render-engine) as a wrapper for the [`react-markdown`](https://github.com/rexxars/react-markdown). We used it in **Docs**, **Blog**, and **Community** views to customize such Markdown elements as panels, or icons next to external links.

![Features provided by the react-markdown library](./react-markdown.png)

### ClusterAssetGroups

> We even found a place for a chunk of Kubernetes implementation in our frontend. In Kyma, we use our in-house, k8s-based component [Rafter](https://kyma-project.io/docs/components/rafter/#overview-rafter-in-kyma) as a backend mechanism for uploading data for documentation topics and storing them in external buckets located in MinIO storage. Rafter is based on AssetGroup and ClusterAssetGroup custom resources (CRs) - we decided to use their structure on the website to configure the left navigation for [documentation topics](https://kyma-project.io/community/guidelines/content#add-new-documentation-to-the-website-add-new-documentation-to-the-website) in the **Docs** and **Community** views, simply by fetching content from these `.yaml` files. Each [ClusterAssetGroup](https://github.com/kyma-project/kyma/tree/master/resources/core/charts/docs/charts/content-ui/templates) is a separate node in the navigation. The content loader uploads the source documentation for a given topic from the path specified in the ClusterAssetGroup CR under the **filter** path. It then renders the sources in the order (`rafter.kyma-project.io/order: "2"`) and under the name (`displayName: "Service Catalog"`) specified in the CRs. Similarly to the documentation component and the react-markdown library, we use ClusterAssetGroup CRs in this way both on the website and for the documentation on Kyma clusters.

Author: I would change it into "data"

> We even found a place for a chunk of Kubernetes implementation in our frontend. In Kyma, we use our in-house k8s-based component [Rafter](https://kyma-project.io/docs/components/rafter/#overview-rafter-in-kyma) as a backend mechanism for uploading data for documentation topics. Rafter provides ClusterAssetGroup custom resources (CRs) that group assets — we decided to use their structure on the website to configure the left-side navigation for [documentation topics](https://kyma-project.io/community/guidelines/content#add-new-documentation-to-the-website-add-new-documentation-to-the-website) in the **Docs** and **Community** views. We did it simply by fetching the content from these `.yaml` files. Each [ClusterAssetGroup](https://github.com/kyma-project/kyma/tree/master/resources/core/charts/docs/charts/content-ui/templates) is a separate node in the navigation. The content loader uploads the source documentation for a given topic from the path specified in the ClusterAssetGroup CR under the **filter** path. It then renders the sources in the order and under the name specified in the CRs (`rafter.kyma-project.io/order: "2"` and `displayName: "Service Catalog"` respectively). Similarly to the documentation component and the react-markdown library, we use ClusterAssetGroup CRs both on the website and for the documentation on Kyma clusters.

Author: I changed K8s Kubernetes but react-markdown is the library name so I will leave it like this.

> We even found a place for a chunk of Kubernetes implementation in our frontend. In Kyma, we use our in-house k8s-based component [Rafter](https://kyma-project.io/docs/components/rafter/#overview-rafter-in-kyma) as a backend mechanism for uploading data for documentation topics. Rafter provides ClusterAssetGroup custom resources (CRs) that group assets — we decided to use their structure on the website to configure the left-side navigation for [documentation topics](https://kyma-project.io/community/guidelines/content#add-new-documentation-to-the-website-add-new-documentation-to-the-website) in the **Docs** and **Community** views. We did it simply by fetching the content from these `.yaml` files. Each [ClusterAssetGroup](https://github.com/kyma-project/kyma/tree/master/resources/core/charts/docs/charts/content-ui/templates) is a separate node in the navigation. The content loader uploads the source documentation for a given topic from the path specified in the ClusterAssetGroup CR under the **filter** path. It then renders the sources in the order and under the name specified in the CRs (`rafter.kyma-project.io/order: "2"` and `displayName: "Service Catalog"` respectively). Similarly to the documentation component and the react-markdown library, we use ClusterAssetGroup CRs both on the website and for the documentation on Kyma clusters.

Author: I changed it this way -> `that group assets, and we decided to`. I did not want to make a separate sentence out of it as I guess `their` wouldn't be so clear then.

## Views and their sources

![Views on the website](./views.png)

Each view on the website takes its sources from a different repository.

> - **Docs** — the `docs` folder in the `kyma` repository

Author: Same reasoning as above

- **Docs** - `docs` folder in the `kyma` repository
- **Blog** - `content/blog-post` folder in the `website` repository
- **Community** - `community` repository

> - **Roadmap** - `capabilities` folder in the community repository for descriptions of our project areas, and GitHub issues with `Epic` and a given capability's labels for the roadmap details.

Author: Yup, I will remove the dot, I will leave `labels` however as this refers to both `Epic` and capability

> - **Landing page** - `content` folder in the `website` repository, including the banner, and the **Used by** section with Kyma users

Author: yes but these are not full sentences so I would rather remove this dot

> Before the website build, all this content is copied to the `website` repository by the [content loader](https://github.com/kyma-project/website/tree/master/tools/content-loader) - our own TypeScript tool we use for fetching:

Author: To be honest, I'm not sure about this one. If we change the order the separate sentence about the content loader seems sort of detached :/

- Content from various repository sources

> - Details of issues from ZenHub and GitHub APIs

Author: We have changed the approach and started to move away from Zenhub but the website logic has not changed yet. I will probably add a note at the end (somewhere in the contributing area) explaining out plans.

## Building and deployment process

Now that you know all pieces of the puzzle, let's have a look at how they fit together. The diagram and description beneath show the whole website building process triggered after merging a PR.

![Website building flow](./building-process.svg)

1. When you merge your PR to the `kyma/docs`or `community` repositories, the GitHub API sends an event to the Netlify function that triggers the master build.

> **NOTE:** The Netlify function is not triggered for the `website` repository since any changes there automatically trigger the master build.

> 2. The build triggers the content loader to fetch the given repository content, along with related ClusterAssetGroups, and perform initial data serialization and filtering. The content loader then moves the fetched content under the `content` folder in the `website` repository.

Author: copied 👍

3. Gatsby reads this content through the connected plugins and retrieves selected data, such as metadata in docs. It later transforms it into a GraphQL schema that could be pulled by React components.

4. Gatsby uses the [Node API](https://www.gatsbyjs.org/docs/node-apis/) to build particular static HTML sites from React components, using data from GraphQL.

5. Gatsby [optimizes](https://www.gatsbyjs.org/docs/performance/) the static sites to increase their performance.

6. Finally, Netlify deploys the static sites.

## Preview

> Before your merge a PR and it gets published on the Kyma website, you can preview your changes. This way you can see if the formatting of the text is correct, images fit well, and links work as expected. That is possible thanks to the [preview feature](https://kyma-project.io/community/guidelines/content#documentation-preview-documentation-preview) supported by Netlify. It attaches links to autogenerated previews of all website views to your PRs.

Reviewer 3: I'm also wondering if we shouldn't change `supported by Netlify` to `providede by Netlify`, since - if I understand correctly - it's their feature that we use, and `support` brings to mind a third-party feature that they support rather than that. But that may be nitpicking... What do you think?

> Before your merge a PR and it gets published on the Kyma website, you can preview your changes. This way you can see if the formatting of the text is correct, images fit well, and links work as expected. That is possible thanks to the [preview feature](https://kyma-project.io/community/guidelines/content#documentation-preview-documentation-preview) supported by Netlify. It attaches links to autogenerated previews of all website views to your PRs.

Author: but for `autogenerated`, I think it is fine this way --> (link)[https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/a/auto]

We enabled this feature on these Kyma repositories:

- `kyma` for changes in the `docs` folder that contains sources of the official Kyma documentation rendered in the **Docs** view.

- `community` for changes rendered in the **Community** view.

- `website` for changes rendered on the landing page and in the **Blog** view.

Previews are built for PRs containing changes made to any file within these repositories, for both successful and failed builds. The only exception is the `kyma` repository where Netlify only builds previews for changes in the `docs` folder and publishes notifications only for successful builds on such PRs.

> When it comes to the building process, it looks very similar to the general flow. The only difference is that the Netlify function is not involved. Every commit on a PR triggers the content loader straightaway, generating only the preview of the view you actually modify.

Reviewer 3: ```suggestion
As for the building process, it looks very similar to the general flow. The only difference is that the Netlify function is not involved. Every commit on a PR triggers the content loader straightaway, generating only the preview of the view you actually modify.
```
This is very preferential, please ignore it if you'd rather stick to your version.

> When it comes to the building process, it looks very similar to the general flow. The only difference is that the Netlify function is not involved. Every commit on a PR triggers the content loader straightaway, generating only the preview of the view you actually modify.

Reviewer 3: `are currently modifying` or `are currently working on`, I agree.

> When it comes to the building process, it looks very similar to the general flow. The only difference is that the Netlify function is not involved. Every commit on a PR triggers the content loader straightaway, generating only the preview of the view you actually modify.

Author: I actually meant `actually` here ;) but if that is misleading, I can give currently as one way or another that is actually the current view you modify ;)

![Successful preview build on a pull request](./successful-preview.png)

## How and where to contribute

> When it comes to our future plans concerning the website, we have some ideas on how to improve its overall performance, simplify contribution and introduce easy feedback options. We log all those ideas as [GitHub issues](https://github.com/kyma-project/website/issues) in the `website` repository - feel encouraged to do the same if an idea for improvement crosses your mind.

Reviewer 3: I'd also change `When it comes to` to `Regarding`, but it's might just be me. I just feel like `when it comes to` has a slightly different ring to it.

> When it comes to our future plans concerning the website, we have some ideas on how to improve its overall performance, simplify contribution and introduce easy feedback options. We log all those ideas as [GitHub issues](https://github.com/kyma-project/website/issues) in the `website` repository - feel encouraged to do the same if an idea for improvement crosses your mind.

Author: I will leave it for now as it is if you don't mind

> You can also move straightaway to action and contribute on your own to the look and feel of https://kyma-project.io/.

Reviewer 3: ```suggestion
You can also jump straight to action and contribute yourself to the look and feel of https://kyma-project.io/.
```

Also, do we want to format the URL as code?

Author: I just changed `contribute yourself` to `contribute to the website yourself`

Author: All done, but. I did not add `yourself`, just left `contribute`.

The contribution flow is quite simple:

> 1. Fork a repo.

Reviewer 4: Can we add a link to our [gitflow](https://kyma-project.io/community/contributing#git-workflow-git-workflow) process? Are we recommending the same process?

Author: Yes, that's the one - I will link to it.

2. Create a PR.
3. Add content and wait for our review and approval.

> The fun may begin when you try to figure out where the sources of all views are. However, we hope this table clarifies it a bit and will help you to find your way through our repositories and the website structure:

Reviewer 3: ```suggestion
The fun may begin when you try to figure out where we keep the sources for all the views. However, we hope this table clarifies it a bit and will help you to find your way through our repositories and the website structure:
```

Or maybe something like:
```It's pretty straightforward until you try to figure out where we keep all the sources for all the views. However, fear not. The table below clarifies it all. We hope it will help you to find your way through our repositories and the website structure.```
to shorten and simplify the sentences a bit.
What do you think?

> The fun may begin when you try to figure out where the sources of all views are. However, we hope this table clarifies it a bit and will help you to find your way through our repositories and the website structure:

Author: I change this section a bit, combines your ideas and added sth from me as well - take a look if that is fine :)

> | Website view | Where? | What? | How? |

Reviewer 3: I'd get rid of the questions mark in the table header for the clean look, but again, that's preferential.

|---|---|---|---|

> | **Landing page** | [`website/content/adopters/adopters.yaml`](https://github.com/kyma-project/website/blob/master/content/adopters/adopters.yaml) | Kyma User | Follow the [instruction](https://github.com/kyma-project/website/blob/master/docs/add-user.md) or [log and issue](https://github.com/kyma-project/website/issues/new?template=new-user-request.md) and we will add it for you. |

Reviewer 3: ```suggestion
| **Landing page** | [`website/content/adopters/adopters.yaml`](https://github.com/kyma-project/website/blob/master/content/adopters/adopters.yaml) | Kyma Adopters | Follow this [instruction](https://github.com/kyma-project/website/blob/master/docs/add-user.md) or [log an issue](https://github.com/kyma-project/website/issues/new?template=new-user-request.md) and we will add it for you. |
```

> | **Landing page** | [`website/content/adopters/adopters.yaml`](https://github.com/kyma-project/website/blob/master/content/adopters/adopters.yaml) | Kyma User | Follow the [instruction](https://github.com/kyma-project/website/blob/master/docs/add-user.md) or [log and issue](https://github.com/kyma-project/website/issues/new?template=new-user-request.md) and we will add it for you. |

Reviewer 3: I'm also for `Kyma Adopters`.

> | **Landing page** | [`website/content/adopters/adopters.yaml`](https://github.com/kyma-project/website/blob/master/content/adopters/adopters.yaml) | Kyma User | Follow the [instruction](https://github.com/kyma-project/website/blob/master/docs/add-user.md) or [log and issue](https://github.com/kyma-project/website/issues/new?template=new-user-request.md) and we will add it for you. |

Author: I changed users to adopters 👍

| **Landing page** | [`website/content/banner/slides.yml`](https://github.com/kyma-project/website/blob/master/content/banner/slides.yml) | Banner | Follow this [instruction](https://github.com/kyma-project/website/blob/master/docs/banner-modification.md). |

> | **Docs** | [`kyma/docs`](https://github.com/kyma-project/kyma/tree/master/docs) | Document or topic | Add a document that follows one of the [templates](https://kyma-project.io/community/guidelines/templates/#document-types-templates-document-types-templates-document-types-for-kyma-components) or follow instructions to [add a new topic](https://kyma-project.io/community/guidelines/content/#add-new-documentation-to-the-website-add-new-documentation-to-the-website). |

Reviewer 2: ```suggestion
| **Docs** | [`kyma/docs`](https://github.com/kyma-project/kyma/tree/master/docs) | Document or topic | Add a document that follows one of the [templates](https://kyma-project.io/community/guidelines/templates/#document-types-templates-document-types-templates-document-types-for-kyma-components) or follow the instructions to [add a new topic](https://kyma-project.io/community/guidelines/content/#add-new-documentation-to-the-website-add-new-documentation-to-the-website). |
```

> | **Blogs** | [`website/content/blog-post`](https://github.com/kyma-project/website/tree/master/content/blog-posts) | Blog post | Follow the [instruction](https://github.com/kyma-project/website/blob/master/docs/write-blog-posts.md). |

Reviewer 3: ```suggestion
| **Blogs** | [`website/content/blog-post`](https://github.com/kyma-project/website/tree/master/content/blog-posts) | Blog post | Follow this [instruction](https://github.com/kyma-project/website/blob/master/docs/write-blog-posts.md). |
```
As @bszwarc changed it in the entry above to make all the entries uniform, I did it here too to keep it uniform.

Author: `I actually unified it the other way round - changed to "the" everywhere.

| **Community** | [`community`](https://github.com/kyma-project/community) | Document | Create a PR. |

> | **Roadmap** | All repos | Epic | Log and issue, assign it to the proper milestone on Zenhub, and add your capability's and `Epic` labels. |

Reviewer 3: Since we're moving away from ZenHub Milestones as we used them before and now use the Milestones to mark the release, should we keep the `assign it to the proper milestone on Zenhub`? Unless you mean the release.

```suggestion
| **Roadmap** | All repos | Epic | Log and issue, assign it to the proper milestone on Zenhub, and add your [capability](https://github.com/kyma-project/community/tree/master/capabilities) and `Epic` labels. |
```
or ```add the `Epic` label and the label for your capability```.

Author: I changed `assign it to the proper milestone on Zenhub` to `assign it to the proper release on GitHub` .

![Website building flow](./building-process.svg)

1. When you merge your PR to the `kyma/docs`or `community` repositories, the GitHub API sends an event to the Netlify function that triggers the master build.

> **NOTE:** The Netlify function is not triggered for the `website` repository since any changes there automatically trigger the master build.

2. The build triggers the content loader to fetch the given repository content, along with related ClusterAssetGroups, and perform initial data serialization and filtering. The content loader then moves the fetched content under the `content` folder in the `website` repository.

3. Gatsby reads this content through the connected plugins and retrieves selected data, such as metadata in docs. It later transforms it into a GraphQL schema that could be pulled by React components.

> 4. Gatsby uses the [Node API](https://www.gatsbyjs.org/docs/node-apis/) to build particular static HTLM sites from React components, using data from GraphQL.

Reviewer: `HTML` :)
